### DISCUSSION TITLE
Last working config

### DISCUSSION THREAD

**Discussion by FrigateEnthusiast** [blakeblackshear/frigate#8891] [Ideas]
We've been all there... make a change or two to your config, restart Frigate and IT BOMBS, somebody is unhappy! 

"What did I change and need to revert or back out??"

I recently made a change to my object list ("This is "safe" change, no need to copy my current config..."), saved with no yml editor grip messages and then my Frigate container would not start, caught in a docker restart loop.

Wasn't my first rodeo so I've learned to copy the current config.yml to last_know_working_config.yml before restarting, AND keep my config directory outside of the container/image, so I could quickly fix/revert the bad config outside of Frigate, and get back up and running. 

What would be a great new feature would be for the Frigate to do a lot of that automatically. Save the current config to <changedatetime> before any edit saves. Then there would always be a history of changes and you could pick a prior and get back up and running. And once Frigate gets to a fully up and running state, it copies the current config to "last_working_config.yml".

Best and most elegant would be if a git repository was detected in the config folder, a git commit (git part of the Frigate image) was performed on each config edit save.

**Comment by banthungprong** [comment]
This will only work if you edit your config in the frigate editor. I do not use the internal editor.
Like user administration it isn't an essential feature of a NVR with realtime object detection.
If you edit your configs with Visual Studio you will have a lot of options to backup previous versions - working perfectly (I use this for all my 3D-OpenSCAD-Files with a git repository). No real need to implement a version control system in frigate.

**Reply by FrigateEnthusiast** [reply]
How do you set that up? I run Visual Studio on a Window's desktop. Wouldn't that require always having the Linux Frigate config directory available as a network location, accessible to the Windows PC? That seems like that can get dodgy...

I assume you keep the configs in a Windows folder and just copy to the Frigate container location, or something like that. But that requires a copy to, for each edit.

**Reply by banthungprong** [reply]
A mounted network location isn't "dodgy" but simply stable. And with plugins you can even use VSCode directly for remote editing the files. (I used this a few years ago with VS on Windows connecting to files on a raspberry).

No I do not copy for each edit - see above explanation.

**Comment by NickM-27** [comment]
The built-in config editor has config validation before it allows saving. If you're not using it then the burden is on you to not mess your config up, I don't see a reason for frigate to do this when it already has a config editor that will make sure you can't save a broken config.

**Reply by FrigateEnthusiast** [reply]
"I recently made a change to my object list ("This is "safe" change, no need to copy my current config..."), **saved with no yml editor grip messages** and then my Frigate container would not start, caught in a docker restart loop."

I have made changes where the editor did not flag anything and yet the instance would not start. A valid and correct format does mean the Frigate instance will always run without blowing up.

**Reply by NickM-27** [reply]
Then that would be a bug with the config editor that should be reported. When the editor saves the config it does the exact same logic that frigate does on startup and if it does not open successfully then the config won't be saved. 

I have never had it save and then not run personally so specifics on what didn't get caught would be good to see

**Reply by NickM-27** [reply]
and to be clear camera issues don't count for that, frigate does not ensure your cameras will actually connect before restarting

**Reply by FrigateEnthusiast** [reply]
I never restart if I get config errors on the save. I'll try to reproduce the fail where all I did was copy and paste all of the possible objects from the doc, just to see what a new outside cam would detect if all were configured in. No errors and yet that attempt was stuck  in a start loop.

Typically for me if a camera config is bad, Frigate will start/run, but just that one camera will not be available.
